### DISCUSSION TITLE
How do I get a stripped absolute path?

### DISCUSSION THREAD

**Discussion by xq114** [xmake-io/xmake#2111] [Q&A]
例如，现有一个path `C:\Users\xq114\src\abc\..\..\proj`，xmake中有什么办法可以把它转化为最简形式`C:\Users\xq114\proj`吗？

**Comment by xq114** [comment]
从实用角度考虑path.absolute应该返回一个最简的path的，但实际上并不会：
```
> xmake l path.absolute C:\Users\xq114\src\abc\..\..\proj
"C:\Users\xq114\src\abc\..\..\proj"
```
但在把相对路径转成绝对路径的时候表现又不一样了
```
PS C:\Users\xq114> xmake l path.absolute src\abc\..\..\proj               
"C:\Users\xq114\proj"
```
这个行为是一个bug吗

**Reply by waruqi** [reply]
不算是 bug，因为还是有效的绝对路径，只不过目前一直没空去处理路径简化，用到现在一直能用，就凑合着这么用了而已。

要简化的话，得在 c 层 translate 里面处理 ，要改进 tb_path_translate 实现才行，目前只做了最基础的  /// 等路径分隔符去重

**Reply by xq114** [reply]
能不能提取出一个额外的路径简化函数？我看tb_path_absolute里面是有这块代码的
https://github.com/tboox/tbox/blob/HEAD/src/tbox/platform/path.c#L226-L227

**Reply by waruqi** [reply]
> 能不能提取出一个额外的路径简化函数？我看tb_path_absolute里面是有这块代码的 https://github.com/tboox/tbox/blob/HEAD/src/tbox/platform/path.c#L226-L227

那块只是做 absolute 但是对 ../../ 这种没做简化，这块应该是 path.translate 也就是 tb_path_translate 干的事，专门用来处理路径简化和格式化的

所以没必要单独再去搞的函数。直接改进这个就行了

**Reply by waruqi** [reply]
absolute 里面只对 /xx/yy + ../../yy 这种对接点的 ../../做了简化

而左右两部分如果中间部分本身存在 ../../，比如 /xx/yy/../../hhh/hg + hhh 这种就直接拼接了 没去简化他们

**Reply by waruqi** [reply]
还有 ./././ 这种也没去简化，目前

**Comment by xq114** [comment]
疑似因为path.absolute没最简化，导致path.relative也出问题
```
PS C:\Users\xq114> xmake l path.relative C:\proj\aa\bb C:\proj              
"aa\bb"
PS C:\Users\xq114> xmake l path.relative C:\proj\aa\bb C:\src\abc\..\..\proj
"C:\proj\aa\bb"
```

**Reply by waruqi** [reply]
加上支持了，见：https://github.com/xmake-io/xmake/pull/2137
